status/api/provider.nim

```diff
@@ -23,3 +28,59 @@ proc callRpc*(self: StatusObject, rpcMethod: string, params: JsonNode):
     return CallRpcResult.ok await self.web3.callRpc(rpcMethod, params)
   except CatchableError as e:
     return CallRpcResult.err e.msg
+
+proc signTransaction(tr: var Transaction, pk: PrivateKey) =
+  let h = tr.txHashNoSignature
+  let s = sign(pk, SkMessage(h.data))
+
+  var r = toRaw(s)
+  let v = r[64]
+
+  tr.R = fromBytesBE(UInt256, r.toOpenArray(0, 31))
+  tr.S = fromBytesBE(UInt256, r.toOpenArray(32, 63))
+
+  tr.V = int64(v)
+
+  if tr.txType == TxType.TxLegacy:
+    tr.V += 27 # TODO! Complete this
+
+proc sendTransaction*(self: StatusObject, fromAddress: EthAddress, transaction: Transaction, password: string, dir: string):
+  Future[SendTransactionResult] {.async.} =
+  if not self.isLoggedIn:
+    return SendTransactionResult.err "Not logged in. Must be logged in"
+
+  try:
+    let accounts = self.userDb.getWalletAccounts()
+    let account = accounts.filter(x => x.address == Address(fromAddress))
+    if account.len == 0:
+      return SendTransactionResult.err "Wallet address not found"
+
+    let loadAccountResult = self.accountsGenerator.loadAccount(account[0].address, password,dir)
+    if loadAccountResult.isErr:
+      return SendTransactionResult.err loadAccountResult.error
+    
+    let accountResult = self.accountsGenerator.findAccount(loadAccountResult.get.id)
+    if accountResult.isErr:
+      return SendTransactionResult.err accountResult.error
+    
+    let
+      settings = self.userDb.getSettings()
+      network = settings.getCurrentNetwork()
+
+    var trx = Transaction(
+      txType: transaction.txType,
+      to: transaction.to,
+      value: transaction.value,
+      maxPriorityFee: transaction.maxPriorityFee,
+      maxFee: transaction.maxFee,
+      gasLimit: transaction.gasLimit,
+      payload: transaction.payload,
+      nonce: transaction.nonce,
+      chainId: ChainId(network.get.config.networkId)
+    )
```

Comment:
Why the need to rebuild the `Transaction` here? Could we instead accept a mutable `Transaction` object in the signature?

Reply:
I tried using `var` but the compiler complained: `Error: invalid type: 'var Transaction' in this context: 'ref SendTransactionTaskArg:ObjectType' for var` . Not sure how to fix it

Reply:
@michaelsbradleyjr, I believe this has something to do with either the task macro not allowing `var` arg types or `json_serialization` not being able to decode it properly (or a combination of both). What do you think?

Reply:
@richard-ramos, easiest to simply create a mutable copy like so. Can also remove the settings/network retrieval above it.

```suggestion
    var trx = transaction
```

Reply:
Nice catch, @emizzle. Afaict, the `type Transaction` we're working with is defined in `vendor/nim-waku/vendor/nim-eth/eth/common/eth_types.nim` using `Transaction* = object`.

So on assignment to `var trx` it gets copied and in the context of the body of `proc sendTransaction` that's enough to satisfy the compiler.

Reply:
Do you know why we can't pass in a mutable `Transaction` object in to the proc?

Reply:
I don't know, not exactly. I'll need to do some experiments tomorrow.